--- stock__-R-4.0.3/src/unix/sys-std.c	2020-06-07 18:15:01.000000000 -0400
+++ patched-R-4.0.3/src/unix/sys-std.c	2021-02-01 14:53:56.521712571 -0500
@@ -30,6 +30,11 @@
 
 /* select() is essential here, but configure has required it */
 
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+
 #ifdef HAVE_CONFIG_H
 # include <config.h>
 #endif
@@ -655,6 +660,28 @@
   }
 }
 
+void write_to_extended_history(char *line){
+    FILE *fptr;
+    long size;
+    char* buf;
+    char* cwdptr;
+    const char* ZCONTEXT    = getenv("ZCONTEXT");
+    size = pathconf(".", _PC_PATH_MAX);
+    buf = (char *)malloc((size_t)size);
+    cwdptr = getcwd(buf, (size_t)size);
+    fptr = fopen(EXTENDED_HISTORY_FILE, "a");
+    fprintf(fptr, ":R:%s:%s:%s:%s: %lu:0;%s\n",
+            MZUNAME,
+            MZHOSTNAME,
+            cwdptr,
+            ZCONTEXT,
+            (unsigned long)time(NULL),
+            line);
+    fclose(fptr);
+    free(buf);
+}
+
+
 static void readline_handler(char *line)
 {
     R_size_t buflen = rl_top->readline_len;
@@ -665,8 +692,10 @@
 	return;
     if (line[0]) {
 # ifdef HAVE_READLINE_HISTORY_H
-	if (strlen(line) && rl_top->readline_addtohistory)
+        if (strlen(line) && rl_top->readline_addtohistory){
+	    write_to_extended_history(line);
 	    add_history(line);
+	}
 # endif
 	/* We need to append a \n if the completed line would fit in the
 	   buffer but not otherwise.  Byte [buflen] is zeroed in
